### DISCUSSION TITLE
#13 - Never ending testing saga 🙌

### DISCUSSION THREAD

**Discussion by AlexMi-Ha** [CUMGroup/AMOGUS#112] [Blog]
![Hello there](https://media4.giphy.com/media/Nx0rz3jtxtEre/giphy.gif?cid=ecf05e473639agwtgpbijpwknfdnvf02dppkrhipi6nt6mrf&ep=v1_gifs_search&rid=giphy.gif&ct=g)
> "Hello There"
> ~ 2005, Space Jesus

Welcome to another sussy blog! 🔥

As already mentioned in last weeks blog, this week we continued with testing! 🔨
We've achieved quite some coverage ^^
![image](https://github.com/CUMGroup/AMOGUS/assets/61502536/b16a681a-628c-472b-b252-55709c747e67)

Every major Service has tests now! ✨

![image](https://github.com/CUMGroup/AMOGUS/assets/61502536/44db1569-4b51-46ed-a036-8febf3825147)
 
Furthermore, we included testing in our Github Action workflows.
When someone opens a Pull Request on our Repo, all Unit Tests get run automatically.
If some tests fail, the pull request gets rejected until all tests succeed!

To review the test results, they can be found In the workflow logs together with the respective code coverage.

![image](https://github.com/CUMGroup/AMOGUS/assets/61502536/9cc60c54-687c-432d-8f2b-db921c7a812f)

Our updated Testplan can be found [here](https://cumgroup.github.io/AMOGUS/Developer%20Docs/Testing/Testplan/).

But that wasn't the only thing we've done this week! 🥳
We also did some quality of life changes in both our Frontend and Backend, including:
- Play Streak now updates after a game session and not at midnight
- Form elements now have input validation directly in the client (Before, only the server validated input).
- UI was refactored a bit to have a better "flow"
- Some pages in the frontend finally have Role-access checks
- Backend service for Leaderboards (stats based on Longest Streak, Current Streak, and Correct/Wrong Ratio)
- Started our first user tests

Additionally, @jas20202, the artist **_among us_**, started work on custom badges rewarded for different streak milestones. ඞ

We are on good track for our first demo release, so don't miss our coming blog posts 👀

Yours truly,
CUMGroup

**Comment by vans1646** [comment]
Hi CUMGroup,
that's really amazing how much coverage you already have! Good job! It's also prodigious that you've already tested every major service, wow!
I'm also glad to hear that you included it in GitHub Actions so it's all automated.

There's nothing I can give you constructive cristicism about. You did great this week, not only because you tested really much, but also because you kept implementing further features.

Looking forward to your next blog post!

Cheers,
Capyclue

**Reply by Annalytic-programming** [reply]
Howdy @vans1646,

Thanks for your lovely comment ☺🥰 We always strive to do a great job but always with a pinch of meme to put a smile in your face 😊😁

Yours truly,
CUMGroup
